Implement API session & error handling

dex4 opened the pull request:
Implement api error handling

dex4 committed:
Implement SessionInterceptor
Implement api error handling

greenfodor commented on app/src/main/java/com/pose/move/feature/auth/login/LoginViewModel.kt:
```diff
@@ -14,7 +14,7 @@ class LoginViewModel @Inject constructor(
     fun login(email: String, password: String) {
         viewModelScope.launch {
-            internalStorageManager.setIsUserLoggedIn(true)
+            internalStorageManager.setAuthToken("")
```
Shouldn't you set a non-empty value here? Right now you're setting it to empty string when you login and when you clear the storage. This way you can't tell if the user is logged in or not, even if it's just for testing purposes

dex4 replied:
Correct, this is WIP code and it should have been marked with a TODO

greenfodor commented on app/src/main/java/com/pose/move/feature/splash/SplashViewModel.kt:
```diff
@@ -12,7 +12,7 @@ class SplashViewModel @Inject constructor(private val internalStorageManager: In
         get() =
             runBlocking {
-                val isUserLoggedIn = internalStorageManager.isUserLoggedIn.first()
+                val isUserLoggedIn = internalStorageManager.authenticationToken.first().isNotEmpty()
```
this is the issue I was talking about in my other comment

greenfodor commented on app/src/main/java/com/pose/move/network/ApiResponse.kt:
```diff
@@ -0,0 +1,22 @@
+
+        object NetworkError : Error
+    }
+}
```
this one's really satisfying ngl 😍

tapp-hadaralex commented on app/src/main/java/com/pose/move/data/preference/AuthTokenProvider.kt:
```diff
@@ -0,0 +1,6 @@
+
+interface AuthTokenProvider {
+
+    fun getToken(): String
```
I'd suggest making it suspend. It is usually blocking i/o call.
Since you have only 1 function, you could convert this into a SAM (not necessary, only if you wish to do so)

tapp-hadaralex commented on app/src/main/java/com/pose/move/feature/auth/register/RegisterEvent.kt:
```diff
@@ -9,4 +9,6 @@ sealed class RegisterEvent {
 
     object Register : RegisterEvent()
+
+    object ClearError : RegisterEvent()
```
You could convert the `RegisterEvent` class into an `sealed interface` instead of class, since you don't take any values in the constructor.

tapp-hadaralex commented on app/src/main/java/com/pose/move/feature/auth/register/RegisterViewModel.kt:
```diff
@@ -25,25 +25,31 @@ class RegisterViewModel @Inject constructor(
-                    _uiState.value.userName,
-                    _uiState.value.password
-                )
+            val result = userRemoteDataSource.registerUser(
```
I would suggest moving all the logic for registering an user into an `*Usecase` class.
That should also be the place where you store your `authToken` and do other logic related to register/login.

ViewModels should be the dumbest components in a project. Their only purpose should be to forward events and responses.

dex4 replied:
I'll solve the useCases issue in the next PR

tapp-hadaralex commented on app/src/main/java/com/pose/move/feature/profile/menu/MenuViewModel.kt:
```diff
@@ -21,7 +21,7 @@ class MenuViewModel @Inject constructor(
                 internalStorageManager.setHasUserAddedLicensePhoto(false)
             }
-            internalStorageManager.setIsUserLoggedIn(false)
+            internalStorageManager.setAuthToken("")
```
I'd suggest moving this into an usecase as well and not scoping it to the viewModel.
You should use a app-wide scope for this (like GlobalScope.launch, but i'd suggest a custom one).
If you navigate away from this, it will get cancelled.

tapp-hadaralex commented on app/src/main/java/com/pose/move/network/ApiOperationWrapper.kt:
```diff
@@ -0,0 +1,42 @@
+    val errorJsonAdapter: ErrorResponseJsonAdapter = ErrorResponseJsonAdapter(moshi)
+
+    suspend inline operator fun <T> invoke(
+        coroutineContext: CoroutineContext = IO,
```
Hide the kotlin's Coroutine Dispatchers behind your own implementation.

You could have an interface like
```
interface MoveDispacters {
    val io: CoroutineContext
    val default: CoroutineContext
    val main: CoroutineContext
}
```

And inject it everywhere. This way, you don't depend directly on the platform, but rather abstract your whole implementation. This also means that if you ever want to change a dispatcher, it'll be very easy to do so.

tapp-hadaralex replied:
I remembered that you could also use Annotations to tell Hilt which Dispatchers to inject. 
Either that or use the above solution

tapp-hadaralex commented on app/src/main/java/com/pose/move/network/ApiOperationWrapper.kt:
```diff
@@ -0,0 +1,42 @@
+import retrofit2.HttpException
+import kotlin.coroutines.CoroutineContext
+
+class ApiOperationWrapper @Inject constructor(moshi: Moshi) {
```
I'd suggest moving this into an Interceptor. Same logic, but you won't have to explicitly call it every time you do a network operation.

The only part that can be moved is the `withContext` one, but even so, the whole ErrorWrapping mechanism could be easily moved into an interceptor.

dex4 committed:
Move API error handling to ErrorInterceptor using project specific exceptions
Map project specific exceptions to ApiOperationResult.Error types
Remove mock services and implement MockServer to handle local calls using an interceptor

dex4 committed:
Implement AuthTokenProvider as suspend
